components/client/src/main/java/com/hotels/styx/client/loadbalancing/strategies/PowerOfTwoStrategy.java

```diff
@@ -0,0 +1,140 @@
+import com.hotels.styx.api.client.ActiveOrigins;
+import com.hotels.styx.api.client.ConnectionPool;
+import com.hotels.styx.api.client.RemoteHost;
+import com.hotels.styx.api.client.loadbalancing.spi.LoadBalancingStrategy;
+import com.hotels.styx.api.client.loadbalancing.spi.LoadBalancingStrategyFactory;
+import com.hotels.styx.api.configuration.Configuration;
+
+import java.util.List;
+import java.util.Random;
+
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.StreamSupport.stream;
+
+
+/**
+ * A Power Of Two load balancing strategy.
+ *
+ * Randomly choose two origins, and return the one with less ongoing requests.
+ */
+public class PowerOfTwoStrategy implements LoadBalancingStrategy {
+    private ActiveOrigins activeOrigins;
+    private Random rng;
+
+    public PowerOfTwoStrategy(ActiveOrigins activeOrigins) {
+        this(activeOrigins, new Random());
+    }
+
+    @VisibleForTesting
+    PowerOfTwoStrategy(ActiveOrigins activeOrigins, Random rng) {
+        this.activeOrigins = activeOrigins;
+        this.rng = rng;
+    }
+
+    /**
+     * A load balancing strategy that favours the origin with the least response time.
+     */
+    public static class Factory implements LoadBalancingStrategyFactory {
+        @Override
+        public LoadBalancingStrategy create(Environment environment, Configuration strategyConfiguration, ActiveOrigins activeOrigins) {
+            return new PowerOfTwoStrategy(activeOrigins);
+        }
+    }
+
+    @Override
+    public Iterable<RemoteHost> vote(Context context) {
+        List<ConnectionPoolStatus> poolsList = stream(activeOrigins.snapshot().spliterator(), false)
+                .map(host -> new ConnectionPoolStatus(host, context))
+                .collect(toList());
+
+        if (poolsList.size() == 0) {
+            return ImmutableList.of();
+        }
+
+        if (poolsList.size() == 1) {
+            return ImmutableList.of(poolsList.get(0).host());
+        }
+
+        ConnectionPoolStatus poolStatus1 = choose(poolsList);
+        ConnectionPoolStatus poolStatus2 = choose(poolsList, poolStatus1);
+
+        return betterOf(poolStatus1, poolStatus2);
+    }
+
+    private Iterable<RemoteHost> betterOf(ConnectionPoolStatus poolStatus1, ConnectionPoolStatus poolStatus2) {
+        int metric1 = poolStatus1.busyConnectionCount() + poolStatus1.pendingConnectionCount();
+        int metric2 = poolStatus2.busyConnectionCount() + poolStatus2.pendingConnectionCount();
+
+        if (metric1 < metric2) {
+            return ImmutableList.of(poolStatus1.host(), poolStatus2.host());
+        } else {
+            return ImmutableList.of(poolStatus2.host(), poolStatus1.host());
+        }
+    }
+
+    private ConnectionPoolStatus choose(List<ConnectionPoolStatus> poolsList, ConnectionPoolStatus another) {
+        int i = 0;
+        ConnectionPoolStatus second = choose(poolsList);
+
+        while (i < 20 && another.host().id() == second.host.id()) {
```

Comment:
This seems a little arbitrary. Why iterate 20 times? Do we need this trial and error approach? It seems like there is no need for this repeating trial-and-error approach, we simply need to exclude the "another" ID when we `choose`.

Some pseudo-code:
```
int indexOfOther = find(pool in pools)
int randomIndex = random(pools.size)
if(randomIndex >= indexOfOther) {
  randomIndex++
}
```